rclcpp/include/rclcpp/init_options.hpp

```diff
@@ -80,11 +81,28 @@ class InitOptions
   const rcl_init_options_t *
   get_rcl_init_options() const;
 
+  /// Retrieve default domain id and set.
+  RCLCPP_PUBLIC
+  void
+  use_default_domain_id();
+
+  /// Set the domain id.
+  RCLCPP_PUBLIC
+  void
+  set_domain_id(size_t domain_id);
+
+  /// Return domain id.
+  RCLCPP_PUBLIC
+  size_t
+  get_domain_id() const;
+
 protected:
   void
   finalize_init_options();
 
 private:
+  // This mutex is recursive so that the operator can ensure atomicity
```

Comment:
right, fix it.

Reply:
> there is a chance that mutex will be locked in InitOptions::finalize_init_options being already locked in assignment operator.

This is only because how the copy assignment operator is implemented.
You can add a private `InitOptions::finalize_init_options_impl` or a free function `finalize_init_options_impl(rcl_init_options_t *)` that doesn't take the mutex.

A `std::recursive_mutex` should only be used when the user can actually call some of the methods that are taking the mutex recursively.

Reply:
that makes sense to me, thanks. will fix is as you suggested.

Reply:
This has been addressed with https://github.com/ros2/rclcpp/pull/1165/commits/e8fb34ac5bd9ad111d9e9b9a584af12f4cbcf092